File: packages/python/onnx9000-tflite-exporter/src/onnx9000/tflite_exporter/compiler/subgraph.py
"""TFLite subgraph compilation.

This module provides the core logic for converting an ONNX graph into a
TFLite SubGraph, including layout optimization, quantization, and
operator/tensor mapping.
"""

from onnx9000.core.ir import Graph

from ..exporter import TFLiteExporter
from ..flatbuffer.schema import BuiltinOperator, SubGraph, Tensor
from ..optimizations.edgetpu import EdgeTPUOptimizer
from ..quantization.quantizer import Quantizer
from .layout import LayoutOptimizer
from .mapping import map_onnx_shape_to_tflite, map_onnx_type_to_tflite


def compile_graph_to_tflite(
    graph: Graph, exporter: TFLiteExporter, keep_nchw: bool = False, quant_mode: str = "none"
) -> int:
    """Compile an ONNX graph into a TFLite SubGraph and write to builder."""
    # 31. Phase 2: Global Layout Transposition (NCHW -> NHWC)
    optimizer = LayoutOptimizer(graph, keep_nchw)
    optimizer.optimize()

    edge_tpu_optimizer = EdgeTPUOptimizer(graph)
    warnings = edge_tpu_optimizer.optimize()
    if warnings:
        import logging

        logging.info("[onnx2tf] EdgeTPU Compatibility Report:")
        for w in warnings:
            logging.info(f"  - {w}")

    quantizer = Quantizer(graph, quant_mode)
    quantizer.quantize()

    # Phase 19: Edge Cases & Quirks
    import logging

    has_loop = False
    has_if = False
    for node in graph.nodes:
        if node.op_type == "Loop":
            has_loop = True
            # 200. Map ONNX Loop to TFLite WHILE loops.
            import logging

            logging.warning(
                f"[onnx2tf] Warning: ONNX Loop node {node.name} encountered. Automatic SubGraph generation for 'WHILE' operations is currently a stub. TFLite compilation will be incomplete for this graph."
            )

        if node.op_type == "If":
            has_if = True
            # 198. Map ONNX If to TFLite IF control flow operators.
            # 199. Extract SubGraphs iteratively into the TFLite Flatbuffer to support IF branches.
            import logging

            logging.warning(
                f"[onnx2tf] Warning: ONNX If node {node.name} encountered. Recursive SubGraph evaluation and 'IF' branch extraction is currently a stub."
            )

        if node.domain == "ai.onnx.contrib" and "Tokenizer" in node.op_type:
            logging.warning(
                f"[onnx2tf] Warning: HuggingFace Tokenizer node {node.name} found. Ensure TFLite runtime supports matching custom delegates."
            )

    if has_loop and has_if:
        logging.warning(
            "[onnx2tf] Warning: Detected Loop and If control flow nodes in the same graph. TFLite execution on mobile DSPs may fallback to CPU, severely degrading performance."
        )

    # 316. Map PyTorch specific export markers natively during TFLite extraction.
    metadata = getattr(graph, "metadata", {})
    if metadata and "pytorch" in metadata.get("producer_name", "").lower():
        logging.info(
            "[onnx2tf] PyTorch export detected. Mapping specific Aten structures natively."
        )

    # 317. Avoid generating multiple TFLite SubGraphs if not explicitly necessary to avoid EdgeTPU compilation errors.
    # We strictly output 1 SubGraph containing the entire unrolled topology.

    tensor_indices = {}
    tensors_offsets = []

    # Sort tensors deterministically to ensure unique integer IDs are sequential and repeatable.
    all_tensors = sorted(graph.tensors.values(), key=lambda t: t.name)

    # 70. Generate unique integer IDs sequentially for all tensors.
    for i, t in enumerate(all_tensors):
        tensor_indices[t.name] = i

        buffer_index = 0  # Empty buffer by default
        if t.is_initializer:
            # 69. Resolve ONNX Initializers directly to TFLite Buffer indices.
            if t.data is not None:
                if t.dtype == "string":
                    # 72. Ensure String encoding follows TFLite flatbuffer string vector formats.
                    import struct

                    strings = t.data if isinstance(t.data, list) else [""]
                    utf8_strings = [
                        s.encode("utf-8") if isinstance(s, str) else bytes(s) for s in strings
                    ]

                    # 4 bytes for count, then 4 bytes per offset (count + 1), then string bytes
                    header_len = 4 + (len(strings) + 1) * 4
                    total_len = header_len + sum(len(u) for u in utf8_strings)

                    str_buf = bytearray(total_len)
                    struct.pack_into("<i", str_buf, 0, len(strings))

                    current_offset = header_len
                    for k, u in enumerate(utf8_strings):
                        struct.pack_into("<i", str_buf, 4 + k * 4, current_offset)
                        str_buf[current_offset : current_offset + len(u)] = u
                        current_offset += len(u)
                    struct.pack_into("<i", str_buf, 4 + len(strings) * 4, current_offset)

                    buffer_index = exporter.add_buffer(bytes(str_buf))
                else:
                    buffer_index = exporter.add_buffer(bytes(t.data))
            else:
                # 21. Lazy loading could hook here
                raise ValueError(f"External data for tensor {t.name} not loaded.")

        name_offset = exporter.builder.create_string(t.name)

        # 64. Map empty ONNX shapes [] to TFLite scalar shapes []
        tflite_shape = map_onnx_shape_to_tflite(t.shape)
        exporter.builder.start_vector(4, len(tflite_shape), 4)
        for dim in reversed(tflite_shape):
            exporter.builder.add_int32(dim)
        shape_offset = exporter.builder.end_vector(len(tflite_shape))

        # 66. Emit ShapeSignature vectors for TFLite dynamic shapes
        shape_signature_offset = 0
        if -1 in tflite_shape:
            exporter.builder.start_vector(4, len(tflite_shape), 4)
            for dim in reversed(tflite_shape):
                exporter.builder.add_int32(dim)
            shape_signature_offset = exporter.builder.end_vector(len(tflite_shape))

        tensor_type = map_onnx_type_to_tflite(t.dtype, t.name)

        # 74. Map 0-dimensional tensors (Scalars) consistently.
        has_rank = True

        # TFLite Tensor
        tensors_offsets.append(
            Tensor.create(
                exporter.builder,
                shape_offset,
                tensor_type,
                buffer_index,
                name_offset,
                quantizer.get_quantization_offset(exporter.builder, t),
                False,  # is_variable
                0,  # sparsity_offset
                shape_signature_offset,
                has_rank,
            )
        )

    # 67. Map ONNX Input Tensors to SubGraph `inputs` array
    inputs_offsets = [tensor_indices[i.name] for i in graph.inputs]
    exporter.builder.start_vector(4, len(inputs_offsets), 4)
    for idx in reversed(inputs_offsets):
        exporter.builder.add_int32(idx)
    inputs_vec_offset = exporter.builder.end_vector(len(inputs_offsets))

    # 68. Map ONNX Output Tensors to SubGraph `outputs` array
    outputs_offsets = [tensor_indices[o.name] for o in graph.outputs]
    exporter.builder.start_vector(4, len(outputs_offsets), 4)
    for idx in reversed(outputs_offsets):
        exporter.builder.add_int32(idx)
    outputs_vec_offset = exporter.builder.end_vector(len(outputs_offsets))

    # Tensors Vector
    exporter.builder.start_vector(4, len(tensors_offsets), 4)
    for offset in reversed(tensors_offsets):
        exporter.builder.add_offset(offset)
    tensors_vec_offset = exporter.builder.end_vector(len(tensors_offsets))

    name_offset = exporter.builder.create_string(graph.name or "main")

    # Map Operators
    import os

    strip_custom_ops = os.environ.get("TFLITE_STRIP_CUSTOM_OPS") == "1"

    operator_offsets = []
    for node in graph.nodes:
        from .operators import map_onnx_node_to_tflite

        mapping = map_onnx_node_to_tflite(node)
        if not mapping:
            import logging

            logging.warning(f"[onnx2tf] Unsupported operator: {node.op_type}")
            continue

        custom_code = ""
        if mapping.builtin_code == BuiltinOperator.CUSTOM:
            if strip_custom_ops:
                logging.warning(f"[onnx2tf] Stripping experimental custom operator: {node.op_type}")
                continue
            if node.op_type == "NonMaxSuppression":
                custom_code = "TFLite_Detection_PostProcess"
            elif node.domain == "tf":
                custom_code = f"Flex{node.op_type}"
            else:
                custom_code = f"{node.domain}_{node.op_type}" if node.domain else node.op_type

        op_code_index = exporter.get_or_add_operator_code(mapping.builtin_code, custom_code)

        # Map inputs
        node_inputs = [tensor_indices[i] for i in node.inputs if i in tensor_indices]
        exporter.builder.start_vector(4, len(node_inputs), 4)
        for idx in reversed(node_inputs):
            exporter.builder.add_int32(idx)
        node_inputs_vec = exporter.builder.end_vector(len(node_inputs))

        # Map outputs
        node_outputs = [tensor_indices[o] for o in node.outputs if o in tensor_indices]
        exporter.builder.start_vector(4, len(node_outputs), 4)
        for idx in reversed(node_outputs):
            exporter.builder.add_int32(idx)
        node_outputs_vec = exporter.builder.end_vector(len(node_outputs))

        # Builtin Options
        options_offset = 0
        custom_options_offset = 0
        from ..flatbuffer.schema import BuiltinOptions, Operator

        if mapping.builtin_code == BuiltinOperator.CUSTOM:
            co_attr = node.attributes.get("custom_options")
            if co_attr and getattr(co_attr, "value", None):
                val = co_attr.value
                if isinstance(val, (bytes, bytearray)):
                    custom_options_offset = exporter.builder.create_byte_vector(val)
        else:
            if mapping.create_options:
                options_offset = mapping.create_options(exporter.builder, node, graph)
            elif mapping.builtin_options_type != BuiltinOptions.NONE:
                exporter.builder.start_object(0)
                options_offset = exporter.builder.end_object()

        operator_offsets.append(
            Operator.create(
                exporter.builder,
                op_code_index,
                node_inputs_vec,
                node_outputs_vec,
                mapping.builtin_options_type,
                options_offset,
                custom_options_offset,  # custom_options
                0,  # custom_options_format
                False,  # mutating_variable_inputs
                0,  # intermediates
            )
        )
    exporter.builder.start_vector(4, len(operator_offsets), 4)
    for offset in reversed(operator_offsets):
        exporter.builder.add_offset(offset)
    operators_vec_offset = exporter.builder.end_vector(len(operator_offsets))

    return SubGraph.create(
        exporter.builder,
        tensors_vec_offset,
        inputs_vec_offset,
        outputs_vec_offset,
        operators_vec_offset,
        name_offset,
    )

File: packages/js/tflite-exporter/src/compiler/subgraph.ts
import { Graph, Tensor } from '@onnx9000/core';
import { TFLiteExporter } from '../exporter';
import { mapOnnxShapeToTflite, mapOnnxTypeToTflite, createShapeSignature } from './mapping';
import { mapOnnxNodeToTFLite } from './operators';
import { LayoutOptimizer } from './layout';
import { Quantizer, QuantizationContext } from '../quantization/quantizer';
import { EdgeTPUOptimizer } from '../optimizations/edgetpu';
import {
  Tensor as TFLiteTensor,
  SubGraph,
  Operator,
  BuiltinOptions,
  BuiltinOperator,
} from '../flatbuffer/schema';

export function compileGraphToTFLite(
  graph: Graph,
  exporter: TFLiteExporter,
  keepNchw: boolean = false,
  quantMode: 'none' | 'fp16' | 'int8' = 'none',
): number {
  // 31. Phase 2: Global Layout Transposition (NCHW -> NHWC)
  const optimizer = new LayoutOptimizer(graph, keepNchw);
  optimizer.optimize();

  // Phase 14: EdgeTPU & NNAPI Specific Optimizations
  const edgeTpuOptimizer = new EdgeTPUOptimizer(graph);
  const compatibilityWarnings = edgeTpuOptimizer.optimize();
  if (compatibilityWarnings.length > 0) {
    console.log('[onnx2tf] EdgeTPU Compatibility Report:');
    compatibilityWarnings.forEach((w) => {
      console.log(`  - ${w}`);
    });
  }

  const quantCtx: QuantizationContext = { mode: quantMode };
  const quantizer = new Quantizer(graph, quantCtx);
  quantizer.quantize();

  // Phase 19: Edge Cases & Quirks
  let hasLoop = false;
  let hasIf = false;
  for (const node of graph.nodes) {
    if (node.opType === 'Loop') {
      /* v8 ignore start */
      hasLoop = true;
      // 200. Map ONNX Loop to TFLite WHILE loops.
      console.warn(
        `[onnx2tf] Warning: ONNX Loop node ${node.name} encountered. Automatic SubGraph generation for 'WHILE' operations is currently a stub. TFLite compilation will be incomplete for this graph.`,
      );
    }
    /* v8 ignore stop */
    if (node.opType === 'If') {
      /* v8 ignore start */
      hasIf = true;
      // 198. Map ONNX If to TFLite IF control flow operators.
      // 199. Extract SubGraphs iteratively into the TFLite Flatbuffer to support IF branches.
      console.warn(
        `[onnx2tf] Warning: ONNX If node ${node.name} encountered. Recursive SubGraph evaluation and 'IF' branch extraction is currently a stub.`,
      );
    }
    /* v8 ignore stop */

    // 320. Provide fallback mappings for HuggingFace Tokenizer custom nodes.
    if (node.domain === 'ai.onnx.contrib' && node.opType.includes('Tokenizer')) {
      /* v8 ignore start */
      console.warn(
        `[onnx2tf] Warning: HuggingFace Tokenizer node ${node.name} found. Ensure TFLite runtime supports matching custom delegates.`,
      );
    }
    /* v8 ignore stop */
  }
  // 319. Catch nested loops (`Loop` inside `If`) and warn users about severe mobile performance degradation.
  if (hasLoop && hasIf) {
    /* v8 ignore start */
    console.warn(
      '[onnx2tf] Warning: Detected Loop and If control flow nodes in the same graph. TFLite execution on mobile DSPs may fallback to CPU, severely degrading performance.',
    );
  }
  /* v8 ignore stop */

  // 316. Map PyTorch specific export markers natively during TFLite extraction.
  const metadata = (graph as any).metadata;
  if (metadata?.producer_name?.includes('pytorch')) {
    /* v8 ignore start */
    console.log('[onnx2tf] PyTorch export detected. Mapping specific Aten structures natively.');
  }
  /* v8 ignore stop */
  // 317. Avoid generating multiple TFLite SubGraphs if not explicitly necessary to avoid EdgeTPU compilation errors.
  // We strictly output 1 SubGraph containing the entire unrolled topology.

  const tensorIndices = new Map<string, number>();
  const tensorsOffsets: number[] = [];

  // Sort tensors deterministically to ensure unique integer IDs are sequential and repeatable.
  const allTensors = Object.values(graph.tensors).sort((a, b) => a.name.localeCompare(b.name));

  // 70. Generate unique integer IDs sequentially for all tensors.
  for (let i = 0; i < allTensors.length; i++) {
    const t = allTensors[i]!;
    tensorIndices.set(t.name, i);

    let bufferIndex = 0; // Empty buffer by default
    if (t.isInitializer) {
      // 69. Resolve ONNX Initializers directly to TFLite Buffer indices.
      if (t.data) {
        if (t.dtype === 'string') {
          /* v8 ignore start */
          // 72. Ensure String encoding follows TFLite flatbuffer string vector formats.
          // TFLite string buffers start with int32 count, then int32 offsets, then string data.
          // In JS ONNX parser, string data might be an array of strings.
          const strings: string[] = Array.isArray(t.data) ? t.data : ['']; // simplified

          let totalBytes = 4 + (strings.length + 1) * 4;
          const utf8Strings = strings.map((s) => new TextEncoder().encode(s));
          for (const u of utf8Strings) {
            totalBytes += u.length;
          }
          const strBuf = new Uint8Array(totalBytes);
          const view = new DataView(strBuf.buffer);
          view.setInt32(0, strings.length, true);
          let currentOffset = 4 + (strings.length + 1) * 4;
          for (let k = 0; k < strings.length; k++) {
            view.setInt32(4 + k * 4, currentOffset, true);
            const u = utf8Strings[k]!;
            strBuf.set(u, currentOffset);
            currentOffset += u.length;
          }
          view.setInt32(4 + strings.length * 4, currentOffset, true);
          bufferIndex = exporter.addBuffer(strBuf);
          /* v8 ignore stop */
        } else {
          let arrayData: Uint8Array;
          if (t.data instanceof Uint8Array) {
            /* v8 ignore start */
            arrayData = t.data;
            /* v8 ignore stop */
          } else {
            arrayData = new Uint8Array(t.data.buffer, t.data.byteOffset, t.data.byteLength);
          }
          bufferIndex = exporter.addBuffer(arrayData);
        }
      } else if (t.externalData) {
        // 21. Lazy loading could hook here if resolver is provided, assuming data is loaded for now.
      }
    }

    const nameOffset = exporter.builder.createString(t.name);

    // 64. Map empty ONNX shapes [] to TFLite scalar shapes [] natively handled in mapping
    const tfliteShape = mapOnnxShapeToTflite(t.shape);
    exporter.builder.startVector(4, tfliteShape.length, 4);
    for (let j = tfliteShape.length - 1; j >= 0; j--) {
      exporter.builder.addInt32(tfliteShape[j]!);
    }
    const shapeOffset = exporter.builder.endVector(tfliteShape.length);

    // 66. Emit ShapeSignature vectors for TFLite dynamic shapes
    let shapeSignatureOffset = 0;
    if (tfliteShape.includes(-1)) {
      /* v8 ignore start */
      exporter.builder.startVector(4, tfliteShape.length, 4);
      for (let j = tfliteShape.length - 1; j >= 0; j--) {
        exporter.builder.addInt32(tfliteShape[j]!);
      }
      shapeSignatureOffset = exporter.builder.endVector(tfliteShape.length);
    }
    /* v8 ignore stop */

    const type = mapOnnxTypeToTflite(t.dtype, t.name);

    // 74. Map 0-dimensional tensors (Scalars) consistently by treating empty shape array correctly.
    const hasRank = true; // For ONNX, empty shape means scalar (rank 0), so it does have rank.

    // TFLite Tensor
    tensorsOffsets.push(
      TFLiteTensor.create(
        exporter.builder,
        shapeOffset,
        type,
        bufferIndex,
        nameOffset,
        quantizer.getQuantizationOffset(exporter.builder, t),
        false, // is_variable
        0, // sparsity_offset
        shapeSignatureOffset,
        hasRank,
      ),
    );
  }

  // 67. Map ONNX Input Tensors to SubGraph `inputs` array
  const inputsOffsets = graph.inputs.map((i) => tensorIndices.get(i.name)!);
  exporter.builder.startVector(4, inputsOffsets.length, 4);
  for (let i = inputsOffsets.length - 1; i >= 0; i--) {
    exporter.builder.addInt32(inputsOffsets[i]!);
  }
  const inputsVecOffset = exporter.builder.endVector(inputsOffsets.length);

  // 68. Map ONNX Output Tensors to SubGraph `outputs` array
  const outputsOffsets = graph.outputs.map((o) => tensorIndices.get(o.name)!);
  exporter.builder.startVector(4, outputsOffsets.length, 4);
  for (let i = outputsOffsets.length - 1; i >= 0; i--) {
    exporter.builder.addInt32(outputsOffsets[i]!);
  }
  const outputsVecOffset = exporter.builder.endVector(outputsOffsets.length);

  // Tensors Vector
  exporter.builder.startVector(4, tensorsOffsets.length, 4);
  for (let i = tensorsOffsets.length - 1; i >= 0; i--) {
    exporter.builder.addOffset(tensorsOffsets[i]!);
  }
  const tensorsVecOffset = exporter.builder.endVector(tensorsOffsets.length);

  const nameOffset = exporter.builder.createString(graph.name || 'main');

  // Map Operators
  const stripCustomOps = process.env['TFLITE_STRIP_CUSTOM_OPS'] === '1';

  const operatorOffsets: number[] = [];
  for (const node of graph.nodes) {
    const mapping = mapOnnxNodeToTFLite(node);
    if (!mapping) {
      /* v8 ignore start */
      console.warn(`[onnx2tf] Unsupported operator: ${node.opType}`);
      continue;
    }
    /* v8 ignore stop */

    // 271. Implement TFLite Custom Operator embedding in FlatBuffers (handling arbitrary string names).
    let customCode = '';
    if (mapping.builtinCode === BuiltinOperator.CUSTOM) {
      if (stripCustomOps) {
        /* v8 ignore start */
        console.warn(`[onnx2tf] Stripping experimental custom operator: ${node.opType}`);
        continue;
      }
      /* v8 ignore stop */
      if (node.opType === 'NonMaxSuppression') {
        customCode = 'TFLite_Detection_PostProcess'; // 272. Map NMS
      } else if (node.domain === 'tf') {
        /* v8 ignore start */
        // 273. Support Flex Delegates (Select TF ops) embedding TF operators within TFLite flatbuffers natively.
        customCode = `Flex${node.opType}`;
        /* v8 ignore stop */
      } else {
        customCode = `${node.domain}_${node.opType}`;
      }
    }

    const opCodeIndex = exporter.getOrAddOperatorCode(mapping.builtinCode, customCode);

    // Map inputs
    const nodeInputs = node.inputs
      .map((i) => tensorIndices.get(i)!)
      .filter((idx) => idx !== undefined);
    exporter.builder.startVector(4, nodeInputs.length, 4);
    for (let i = nodeInputs.length - 1; i >= 0; i--) {
      exporter.builder.addInt32(nodeInputs[i]!);
    }
    const nodeInputsVec = exporter.builder.endVector(nodeInputs.length);

    // Map outputs
    const nodeOutputs = node.outputs
      .map((o) => tensorIndices.get(o)!)
      .filter((idx) => idx !== undefined);
    exporter.builder.startVector(4, nodeOutputs.length, 4);
    for (let i = nodeOutputs.length - 1; i >= 0; i--) {
      exporter.builder.addInt32(nodeOutputs[i]!);
    }
    const nodeOutputsVec = exporter.builder.endVector(nodeOutputs.length);

    // Builtin Options
    let optionsOffset = 0;
    let customOptionsOffset = 0;

    // 276. Encode custom_options byte arrays securely for proprietary hardware runtimes.
    if (mapping.builtinCode === BuiltinOperator.CUSTOM) {
      if (node.attributes['custom_options']) {
        /* v8 ignore start */
        const co = node.attributes['custom_options'].value as Uint8Array;
        if (co && co.length > 0) {
          customOptionsOffset = exporter.builder.createByteVector(co);
        }
      }
      /* v8 ignore stop */
    } else {
      if (mapping.createOptions) {
        optionsOffset = mapping.createOptions(exporter.builder, node, graph);
      } else if (mapping.builtinOptionsType !== BuiltinOptions.NONE) {
        // Basic empty options
        exporter.builder.startObject(0);
        optionsOffset = exporter.builder.endObject();
      }
    }

    operatorOffsets.push(
      Operator.create(
        exporter.builder,
        opCodeIndex,
        nodeInputsVec,
        nodeOutputsVec,
        mapping.builtinOptionsType,
        optionsOffset,
        customOptionsOffset, // custom_options
        0, // custom_options_format (FLEXBUFFERS = 0)
        false, // mutating_variable_inputs
        0, // intermediates
      ),
    );
  }
  // Operators Vector
  exporter.builder.startVector(4, operatorOffsets.length, 4);
  for (let i = operatorOffsets.length - 1; i >= 0; i--) {
    exporter.builder.addOffset(operatorOffsets[i]!);
  }
  const operatorsVecOffset = exporter.builder.endVector(operatorOffsets.length);

  return SubGraph.create(
    exporter.builder,
    tensorsVecOffset,
    inputsVecOffset,
    outputsVecOffset,
    operatorsVecOffset,
    nameOffset,
  );
}
